### PR TITLE
Fill the Symfony Response with real responded headers

### PR DESCRIPTION
**Pull request by Timo-Linde**
#33 Automatically map response headers

### CODE REVIEW

File: Services/Curl.php [hunk 1]
```diff
@@ -85,19 +85,23 @@ public function sendRequest($url, $method, array $options = array(), $payload =
...
         $this->curlOptionsHandler->setOption(CURLOPT_RETURNTRANSFER, true);
+        $this->curlOptionsHandler->setOption(CURLOPT_HEADER, true);
 
         $this->setUrl($url);
         $this->setMethod($method);
         $this->setPayload($payload);
         curl_setopt_array($this->curl, $this->curlOptionsHandler->getOptions());
 
         $curlResponse = $this->execute();
+        $headerSize = curl_getinfo($this->curl, CURLINFO_HEADER_SIZE);
```
Comment (TobiasHauck) [inline Services/Curl.php]: Let's make a new class "ResponseHeaders" with a ```public static function create($curlResponse, $headerSize)```.

The create function returns an array containing valid headers (the result of http_parse_headers).
So this is what the code would look like here:

```
$curlResponse = $this->execute();
$curlMetaData = (object) curl_getinfo($this->curl);
$headers     = ResponseHeaders::create($curlResponse, $curlMetaData->CURLINFO_HEADER_SIZE);

$this->curlOptionsHandler->reset();
          function_exists('curl_reset') ? curl_reset($this->curl) : curl_setopt_array($this->curl, $this->curlOptionsHandler->getOptions());

return $this->createResponse($content, $headers, $curlMetaData);
```

File: Services/Curl.php [hunk 2]
```diff
@@ -122,16 +126,17 @@ private function execute() {
...
      * @param \stdClass $curlMetaData
      *
      * @return Response
      */
-    private function createResponse($curlResponse, \stdClass $curlMetaData) {
+    private function createResponse($content, $headers, \stdClass $curlMetaData) {
         $response = new Response();
-        $response->setContent($curlResponse);
+        $response->setContent($content);
+        $response->headers->add(self::httpParseHeaders($headers));
```
Comment (TobiasHauck) [inline Services/Curl.php]: According to my comment before this would just be:
```
$response->headers->add($headers);
```

File: Services/Curl.php [hunk 3]
```diff
@@ -188,4 +193,33 @@ private function getError() {
...
+    /**
+     * Parse Http Headers in array
+     * 
+     * @param string $header
+     * @return Array
+     */
+     public static function httpParseHeaders($header) {
+         if (function_exists('http_parse_headers')) return http_parse_headers($header);
+
+         $retVal = array();
```
Comment (TobiasHauck) [inline Services/Curl.php]: I have not tested the it (the following code is just a sketch), but it should better look like:

```
$fields = explode("\r\n", preg_replace('/\x0D\x0A[\x09\x20]+/', ' ', $header));

return array_reduce($fields, function($carry, $field) {
    $match = [];
    if (!preg_match('/([^:]+): (.+)/m', $field, $match)) return $carry;

    $match[1] = preg_replace_callback('/(?<=^|[\x09\x20\x2D])./',function($matches) {
        return strtoupper($matches[0]);
    }, strtolower(trim($match[1])));

    $carry[$match[1]] = isset($carry[$match[1]]) ? [$carry[$match[1]], $match[2]] : trim($match[2]);
    return $carry;
}, []);
```

File: Tests/Functional/Services/CurlTest.php [hunk 4]
```diff
@@ -102,4 +102,54 @@ public function setContentType() {
         $this->assertInstanceOf('Symfony\Component\HttpFoundation\Response', $response);
         $this->assertSame('application/json', $response->headers->get('Content-Type'));
     }
+
```
Comment (TobiasHauck) [inline Tests/Functional/Services/CurlTest.php]: Do you have an idea how to test both branches of the function? (with function http_parse_headers existing and without)?

**Comment by TobiasHauck** [comment]
Thank you for opening the PR and sorry for responding so late!

File: Tests/Functional/Services/CurlTest.php (hunk 4 above)
Comment (Timo-Linde) [inline Tests/Functional/Services/CurlTest.php]: I think the http_parse_headers function from the pecl extension should not be tested. We can create a second function with the own http_parse_headers function which will be tested. In the main function there is only the if function exists.

File: Services/Curl.php (hunk 3 above)
Comment (Timo-Linde) [inline Services/Curl.php]: I have tested it only with phpunit -> no error so it should work

**Comment by TobiasHauck** [comment]
Just cloned your branch to check the code coverage. ```make test``` errored:
```$content = substr($curlResponse, $headerSize);``` returns ```FALSE``` if the header size is equal to strlen. So what you need to add is ```$content = empty($content) ? '' : $content;```

File: Services/Curl.php [hunk 5]
```diff
@@ -85,19 +85,23 @@ public function sendRequest($url, $method, array $options = array(), $payload =
...
 
         $this->setUrl($url);
         $this->setMethod($method);
         $this->setPayload($payload);
         curl_setopt_array($this->curl, $this->curlOptionsHandler->getOptions());
 
         $curlResponse = $this->execute();
+        $headerSize = curl_getinfo($this->curl, CURLINFO_HEADER_SIZE);
+        $headers = ResponseHeaders::create($curlResponse, $headerSize);
+        $content = substr($curlResponse, $headerSize);
```
Comment (TobiasHauck) [inline Services/Curl.php]: add 
```$content = empty($content) ? '' : $content;```

File: Services/ResponseHeaders.php [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+         return self::httpParseHeadersPeclReplacement($header);
+     }
+     
+    /**
+     * Parse Http Headers in array
+     * 
+     * @param string $header
+     * @return Array
+     */
+     public static function httpParseHeadersPeclReplacement($header) {
```
Comment (TobiasHauck) [inline Services/ResponseHeaders.php]: This one should be private

File: Services/ResponseHeaders.php [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    }
+    
+    /**
+     * Parse Http Headers in array
+     * 
+     * @see https://pecl.php.net/
+     * @param string $header
+     * @return Array
+     */
+     public static function httpParseHeaders($header) {
```
Comment (TobiasHauck) [inline Services/ResponseHeaders.php]: Private function, call just by using ```ResponseHeaders::create```

File: Services/ResponseHeaders.php [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+ * CircleRestClientBundle is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with CircleRestClientBundle.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace Circle\RestClientBundle\Services;
```
Comment (TobiasHauck) [inline Services/ResponseHeaders.php]: We should add a new namespace prefix as this is a kind of "type":
```namespace Circle\RestClientBundle\Types;```

File: Services/ResponseHeaders.php [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+
+/**
+ * Gets the Response Headers from a Curl Response
+ *
+ * @author    Timo Linde <info@timo-linde.de>
+ * @copyright 2016 TeeAge-Beatz UG
+ */
+class ResponseHeaders{
```
Comment (TobiasHauck) [inline Services/ResponseHeaders.php]: ```class ResponseHeaders {```

space between class name and opening brace

File: Services/ResponseHeaders.php [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public static function create($curlResponse, $headerSize) { 
+        $header = substr($curlResponse, 0, $headerSize);
+        
+
```
Comment (TobiasHauck) [inline Services/ResponseHeaders.php]: ```return self::httpParseHeaders(substr($curlResponse, 0, $headerSize));```

File: Services/ResponseHeaders.php (hunk 6 above)
Comment (TobiasHauck) [inline Services/ResponseHeaders.php]: Please move the code of this function to httpParseHeaders as we don't need the seperation any more regarding the suggested create function

File: Services/ResponseHeaders.php [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+     * Parse Http Headers in array
+     * 
+     * @param string $header
+     * @return Array
+     */
+     public static function httpParseHeadersPeclReplacement($header) {
+        $fields = explode("\r\n", preg_replace('/\x0D\x0A[\x09\x20]+/', ' ', $header));
+        
+        if(empty($fields)) {
+
```
Comment (TobiasHauck) [inline Services/ResponseHeaders.php]:
```
if(empty($fields)) return [];
```

File: Tests/Functional/Services/ResponseHeadersTest.php [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+ * @SuppressWarnings("PHPMD.StaticAccess")
+ */
+class ResponseHeadersTest extends \PHPUnit_Framework_TestCase {    
+    
+    /**
+     * @test
+     * @covers ::httpParseHeadersPeclReplacement
+     * @covers ::<private>
+     */
+    public function testHttpParseHeadersPeclReplacement() {
```
Comment (TobiasHauck) [inline Tests/Functional/Services/ResponseHeadersTest.php]: testCreate

File: Tests/Functional/Services/ResponseHeadersTest.php [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+ *
+ * @coversDefaultClass Circle\RestClientBundle\Services\ResponseHeaders
+ *
+ * @SuppressWarnings("PHPMD.StaticAccess")
+ */
+class ResponseHeadersTest extends \PHPUnit_Framework_TestCase {    
+    
+    /**
+     * @test
+     * @covers ::httpParseHeadersPeclReplacement
```
Comment (TobiasHauck) [inline Tests/Functional/Services/ResponseHeadersTest.php]: ```@covers ::create```

File: Tests/Functional/Services/ResponseHeadersTest.php [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+    public function testHttpParseHeadersPeclReplacement() {
+        $headers = "HTTP/1.1 302 Found\r\n".
+            "Date: Tue, 21 Apr 2015 13:57:48 GMT\r\n".
+            "Server: Apache/2.2.22 (Debian)\r\n".
+            "Location: https://test.de\r\n".
+            "Vary: Accept-Encoding\r\n".
+            "Content-Length: 284\r\n".
+            "Content-Type: text/html; charset=iso-8859-1\r\n";
+        
+        $parsedHeader = ResponseHeaders::httpParseHeadersPeclReplacement($headers);
```
Comment (TobiasHauck) [inline Tests/Functional/Services/ResponseHeadersTest.php]:
```
$parsedHeader = ResponseHeaders::create($headers, strlen($headers));
```

File: Tests/Functional/Services/ResponseHeadersTest.php [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+        $this->assertEquals("Accept-Encoding", $parsedHeader["Vary"]);
+        $this->assertEquals("284", $parsedHeader["Content-Length"]);
+        $this->assertEquals("text/html; charset=iso-8859-1", $parsedHeader["Content-Type"]);
+        
+        $headers = "content-type: text/html; charset=UTF-8\r\n".
+            "Server: Funky/1.0\r\n".
+            "Set-Cookie: foo=bar\r\n".
+            "Set-Cookie: baz=quux\r\n".
+            "Folded: works\r\n\ttoo\r\n";
+        $returnValue = ResponseHeaders::httpParseHeadersPeclReplacement($headers);
```
Comment (TobiasHauck) [inline Tests/Functional/Services/ResponseHeadersTest.php]: Please move it to another testcase, let's call it testCreateCookies

Comment (TobiasHauck) [inline Tests/Functional/Services/ResponseHeadersTest.php]:
```
$parsedHeader = ResponseHeaders::create($headers, strlen($headers));
```

**Comment by TobiasHauck** [comment]
May you be so kind and add my suggestions? I think after these changes we can merge the PR. If you give me your email address I will add you to the AUTORS file ;) Thanks again!

File: Services/Curl.php (hunk 5 above)
Comment (TobiasHauck) [inline Services/Curl.php]: Please align all the equal signs

File: Services/Curl.php [hunk 16]
```diff
@@ -122,16 +128,17 @@ private function execute() {
      *
      * @SuppressWarnings("PHPMD.StaticAccess");
      *
-     * @param string    $curlResponse
+     * @param string    $content
+     * @param array    $headers
```
Comment (TobiasHauck) [inline Services/Curl.php]: alignment (1 missing space)

File: Tests/Functional/Types/ResponseHeadersTest.php [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+        $this->assertEquals(count($returnValue['Set-Cookie']), 2);
+    }
+    
+    /**
+     * @test
+     * @covers ::create
+     * @covers ::<private>
+     */
+    public function testCreateFolded() {
+        $headers = "Folded: works\r\n\ttoo\r\n";
```
Comment (TobiasHauck) [inline Tests/Functional/Types/ResponseHeadersTest.php]: alignment

File: Types/ResponseHeaders.php [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+ * @author    Timo Linde <info@timo-linde.de>
+ * @copyright 2016 TeeAge-Beatz UG
+ */
+class ResponseHeaders {
+    /**
+     * Creates Header array from a Response string
+     * 
+     * @see https://pecl.php.net/
+     * @param string $curlResponse
+     * @param string $headerSize
```
Comment (TobiasHauck) [inline Types/ResponseHeaders.php]: alignment

File: Parsers/HTTPHeadersParser.php [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ * @author    Timo Linde <info@timo-linde.de>
+ * @copyright 2016 TeeAge-Beatz UG
+ */
+class HTTPHeadersParser {    
+    /**
+     * Parse Http Headers in array
+     * 
+     * @param string $headers
+     * @return Array
```
Comment (TobiasHauck) [inline Parsers/HTTPHeadersParser.php]: array (lower letters) and alignment

File: Types/ResponseHeaders.php [hunk 20]
```diff
@@ -30,10 +30,10 @@ class ResponseHeaders {
     /**
      * Creates Header array from a Response string
      * 
-     * @see https://pecl.php.net/
-     * @param string $curlResponse
-     * @param string $headerSize
-     * @return Array
+     * @see    https://pecl.php.net/
```
Comment (Timo-Linde) [inline Types/ResponseHeaders.php]: like this?

Comment (TobiasHauck) [inline Types/ResponseHeaders.php]: exactly like this

**Comment by Timo-Linde** [comment]
Can you merge this PR and tag a new version? I need this code for 3 productive applications.

**Comment by Timo-Linde** [comment]
Thank you very much. Keep up the great work. = )

**Comment by TobiasHauck** [comment]
Thank you for contributing! I'll add you to the AUTHORS file very soon!

**Pull request merged**